File: 代码/acwing/src/main/java/base/binarySearch/leet911.java
package base.binarySearch;

import grady.Leet1383;

import java.util.HashMap;
import java.util.Map;

/**
 * @author: Zekun Fu
 * @date: 2022/9/24 11:29
 * @Description:在线选举，区间二分
 *
 *  1. 区间问题
 *  2. 二分
 *  3. 对于
 */
public class leet911 {
    Map<Integer, Integer> mp;
    int[] interval;
    int[] nums;
    int n;

    public leet911(int[] persons, int[] times) {
        n = times.length;
        interval = new int[n];
        nums = new int[n];
        mp = new HashMap<>();
        int[] cnt = new int[6001];
        int pre = 0;
        for (int i = 0; i < n; i++) {
            int p = persons[i];
            int t = times[i];
            cnt[p]++;
            if (cnt[p] >= cnt[pre]) {
                pre = p;
            }
            if (i != n - 1) interval[i] = times[i + 1];
            else interval[i] = Integer.MAX_VALUE;
            nums[i] = pre;
        }
    }

    public int query(int t) {
        int l = 0, r = n;
        // System.out.println(interval[n - 1]);
        while (l < r) {
            int mid = (l + r) >> 1;
            // System.out.println(mid + " " + interval[mid]);
            if (interval[mid] <= t) {
                // System.out.println("in");
                l = mid + 1;
            } else r = mid;
        }
        return nums[l];
    }

    public static void main(String[] args) {
        int[] persons = {0, 0, 0, 0, 1};
        int[] times = {0, 6, 39, 52, 75};
        leet911 le = new leet911(persons, times);
        le.query(68);
    }
}
//}["TopVotedCandidate","q","q","q","q","q","q","q","q","q","q"]
//        [[[0,0,0,0,1],[0,6,39,52,75]],[45],[49],[59],[68],[42],[37],[99],[26],[78],[43]]

File: 代码/leetcodeP/math/leet911.py
"""
@author:fuzekun
@file:leet911.py
@time:2022/09/24
@description:

1. 数学
2. 二分查找
"""

# [0, 5): 0
# [5, 10):1
# [10, 15):1
# [15, 20):0
# [20, 25):0
# [25, 30):1
# [30, inf):0
from collections import defaultdict, Counter
from typing import List


class TopVotedCandidate:
    # [0, 5): 0
    # [5, 10):1
    # [10, 15):1
    # [15, 20):0
    # [20, 25):0
    # [25, 30):1
    # [30, inf):0

    mp = defaultdict(int)
    inteval = []
    nums = []
    n = 0

    def __init__(self, persons: List[int], times: List[int]):
        cnt = Counter()
        pre = 0
        self.n = len(times)
        self.inteval = times
        self.inteval.append(float('inf'))
        for i, (p, t) in enumerate(zip(persons, times)):
            cnt[p] += 1
            if cnt[p] >= cnt[pre]:
                # print(pre, p, cnt[pre], cnt[p])
                pre = p
            self.nums.append(pre)

    def q(self, t: int) -> int:
        l, r = 1, self.n
        # 找到最后一个大于他的右边界
        while l < r:
            mid = l + r >> 1
            if self.inteval[mid] > t:
                r = mid
            else:
                l = mid + 1

        # print(l)
        return self.nums[l - 1]



if __name__ == '__main__':
    s = TopVotedCandidate([0,0,0,0,1], [0,6,39,52,75])
    print(s.q(45))
# Your TopVotedCandidate object will be instantiated and called as such:
# obj = TopVotedCandidate(persons, times)
# param_1 = obj.q(t)